src/pubsub/ua_pubsub_keystorage.h

```diff
@@ -0,0 +1,396 @@
+     * referenced by multiple reader / writer groups have a reference count to manage free
+     */
+    UA_UInt32 referenceCount;
+
+    /**
+     * array of keys. the elements inside this array have a next pointer.
+     * keyList can therefore be used as linked list.
+     */
+    keyListItems keyList;
+
+    /**
+     * size of the keyList
+     */
+    size_t keyListSize;
+
+    /**
+     * The maximum number of Past keys a keystorage is allowed to store
+     */
+    UA_UInt32 maxPastKeyCount;
+
+    /**
+     * The maximum number of Future keys a keyStorage is allowed to store
+     */
+    UA_UInt32 maxFutureKeyCount;
+
+    /*
+     * The maximum keylist size, calculated from maxPastKeyCount and maxFutureKeyCount
+     */
+    UA_UInt32 maxKeyListSize;
+
+    /**
+     * The SecurityTokenId that appears in the header of messages secured with the
+     * CurrentKey. It starts at 1 and is incremented by 1 each time the KeyLifetime
+     * elapses even if no keys are requested. If the CurrentTokenId increments past the
+     * maximum value of UInt32 it restarts a 1.
+     */
+    UA_UInt32 currentTokenId;
+
+    /**
+     *  the current key used to secure the messages
+     */
+    UA_PubSubKeyListItem *currentItem;
+
+    /**
+     * keyLifeTime used to update the CurrentKey from the Local KeyStorage
+     */
+    UA_Duration keyLifeTime;
+
+    /**
+     * id used to register the callback to retrieve the keys related to this security
+     * group
+     */
+    UA_UInt64 callBackId;
+
+    /**
+     * Pointer to the key storage list
+     */
+    LIST_ENTRY(UA_PubSubKeyStorage) keyStorageList;
+
+} UA_PubSubKeyStorage;
+
+/**
+ * @brief Allocates memory for a UA_PubSubKeyStorage and returns a pointer of the
+ * PubSubKeyStorage. The memory must be freed by calling UA_PubSubKeyStorage_delete or
+ * UA_PubSubKeyStorage_removeKeyStorage
+ *
+ * @return UA_PubSubKeyStorage*
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_new(void);
+
+/**
+ * @brief Find the Keystorage from the Server KeyStorageList and returns the pointer to
+ * the keystorage
+ *
+ * @param server holds the keystoragelist
+ * @param securityGroupId of the keystorage to be found
+ * @param keyStorage is the pointer to returned keystorage, if not found null is returned
+ * @return Pointer to the keystorage on success, null pointer on failure
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(UA_Server *server,
+                                                     UA_String securityGroupId);
+
+/**
+ * @brief retreives the security policy pointer from the PubSub configuration by
+ * SecurityPolicyUri
+ *
+ * @param server the server object
+ * @param securityPolicyUri the URI of the security policy
+ * @param policy the pointer to the security policy
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_getPubSubSecurityPolicyByUri(UA_Server *server,
```

Comment:
`UA_Server_findPubSubSecurityPolicy`

Reply:
This can then be reused in other places as well.

Reply:
should it also be public? UA_EXPORT?

Reply:
Changed it to UA_Server_findPubSubSecurityPolicy